File: app/src/main/java/mvp/model/domain/GameManager.java
package mvp.model.domain;

import java.util.Map;
import java.util.WeakHashMap;

import mvp.model.entity.Card;
import mvp.model.entity.Game;
import mvp.model.utils.Constants;
import mvp.view.game.GameContract;

/**
 * Created by Manuel Lorenzo on 21/03/2017.
 */

public class GameManager implements GameDomainContract.Manager {
    private Map<Integer, Card> facingUpCardsWeakHashMap = new WeakHashMap<>(Constants.NUMBER_MATCHING_CARDS);
    private GameContract.Presenter gamePresenter;
    private Game game;

    public GameManager(GameContract.Presenter gamePresenter) {
        this.gamePresenter = gamePresenter;
        this.game = new Game();
    }

    @Override
    public void cardFlipped(int position, Card card) {
        if (!facingUpCardsWeakHashMap.containsKey(position)) {
            facingUpCardsWeakHashMap.put(position, card);
            if (shouldCheckForCardsMatch()) {
                // This would be the last card, we should check for matches.
                String id = card.getId();
                boolean matchFound = true;
                for (Map.Entry<Integer, Card> entry : facingUpCardsWeakHashMap.entrySet()) {
                    Card loopCard = entry.getValue();
                    if (!loopCard.getId().equals(id)) {
                        matchFound = false;
                    }
                }
                if (matchFound) {
                    // Match found; remove cards from dataset
                    gamePresenter.removeViewFlipper();
                    for (Map.Entry<Integer, Card> entry : facingUpCardsWeakHashMap.entrySet()) {
                        Card card1 = entry.getValue();
                        gamePresenter.notifyAdapterItemRemoved(card1.getId());
                    }
                    facingUpCardsWeakHashMap.clear();
                    increaseGameScore();
                    gamePresenter.onGameScoreIncreased(getGameScore());
                    if (isGameFinished()) {
                        gamePresenter.onGameFinished();
                    }
                    gamePresenter.removeViewFlipper();
                } else {
                    // Match not found; turn cards over and start from the beginning.
                    gamePresenter.onTurnCardsOver();
                }
            }
        }
    }

    @Override
    public void removeCardsFromMap() {
        for (Map.Entry<Integer, Card> entry : facingUpCardsWeakHashMap.entrySet()) {
            gamePresenter.notifyAdapterItemChanged(entry.getKey());
        }
        facingUpCardsWeakHashMap.clear();
    }

    @Override
    public boolean shouldDispatchUiEvent() {
        return facingUpCardsWeakHashMap.size() != Constants.NUMBER_MATCHING_CARDS;
    }

    private boolean shouldCheckForCardsMatch() {
        return facingUpCardsWeakHashMap.size() == Constants.NUMBER_MATCHING_CARDS;
    }

    private boolean isGameFinished() {
        return game.getScore() == Integer.valueOf(Constants.CARDS_PER_PAGE) / Constants.NUMBER_MATCHING_CARDS;
    }

    private void increaseGameScore() {
        game.setScore(game.getScore() + 1);
    }

    private int getGameScore() {
        return game.getScore();
    }
}

File: app/src/main/java/kittentrate/ui/game/GameManager.kt
package kittentrate.ui.game

/**
 * Created by Manuel Lorenzo on 21/03/2017.
 */

//class GameManager(private val gameViewModel: GameViewModel) {
//    private val facingUpCardsWeakHashMap = WeakHashMap<Int, Card>(Constants.NUMBER_MATCHING_CARDS)
//    private val gameObservable: Observable<Game> = Observable.just(Game())
//
//    val gameScore: Observable<Int>
//        get() = gameObservable.map { t: Game -> t.score }
//
//    fun cardFlipped(position: Int, card: Card) {
//        if (!facingUpCardsWeakHashMap.containsKey(position)) {
//            facingUpCardsWeakHashMap[position] = card
//            if (shouldCheckForCardsMatch()) {
//                // This would be the last card, we should check for matches.
//                val id = card.id
//                var matchFound = true
//                for ((_, loopCard) in facingUpCardsWeakHashMap) {
//                    if (loopCard.id != id) {
//                        matchFound = false
//                    }
//                }
//                if (matchFound) {
//                    // Match found; remove MATCHED cards from data set
//                    gameViewModel.removeViewFlipper()
//                    for ((_, card1) in facingUpCardsWeakHashMap) {
//                        gameViewModel.notifyAdapterItemRemoved(card1.id)
//                    }
//                    facingUpCardsWeakHashMap.clear()
//                    increaseGameScore()
//                    gameViewModel.removeViewFlipper()
//                } else {
//                    // Match not found; turn cards over and start from the beginning.
//                    decreaseGameScore()
//                    gameViewModel.onTurnCardsOver()
//                }
//                //gameViewModel.onGameScoreChanged(gameScore)
//            }
//        }
//    }
//
//    fun removeCardsFromMap() {
//        for ((key) in facingUpCardsWeakHashMap) {
//            gameViewModel.notifyAdapterItemChanged(key)
//        }
//        facingUpCardsWeakHashMap.clear()
//    }
//
//    fun shouldDispatchUiEvent(): Boolean {
//        return facingUpCardsWeakHashMap.size != Constants.NUMBER_MATCHING_CARDS
//    }
//
//    fun resetScore() {
//        resetGameScore()
//    }
//
//    private fun shouldCheckForCardsMatch(): Boolean {
//        return facingUpCardsWeakHashMap.size == Constants.NUMBER_MATCHING_CARDS
//    }
//
//    private fun increaseGameScore() {
//        gameObservable.map { t: Game -> t.score += 1 }
//    }
//
//    private fun decreaseGameScore() {
//        gameObservable.map { t: Game -> t.score -= 1 }
//    }
//
//    private fun resetGameScore() {
//        gameObservable.map { t: Game -> t.score = 0 }
//    }
//}
